### PR TITLE
Null annotate the GitCommands project

### PR DESCRIPTION
**Pull request by drewnoakes**
Relates to https://github.com/gitextensions/gitextensions/issues/8728

## Proposed changes

- Introduce C# "nullable reference types" annotations to the `GitCommand` project

These are quite broad changes across many files, but they follow very repetitive patterns.

There may be places where we may wish to stop accepting null values (the code is very tolerant of nulls, which isn't great) but for this first pass i wanted to simply capture the _actual_ places we flow `null` values in the current code.

Because we target `net461` we do not have an annotated BCL, hence the new `Strings` class which provides annotations for the two common `string.IsNullOr*` methods.

## Test methodology 

- Manual testing

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by gerhardol** [comment]
Looks fine to me, but hard to formally review.

Did you see any possible bugs fixed by this?
Is this introducing any runtime risks?
How much of this was search&replace and how much required manual changes?
(I could see a few obvious manual touchups but quite a lot seem to require manual too).

**Comment by drewnoakes** [comment]
> Looks fine to me, but hard to formally review.

Yes, hard to review.

> Did you see any possible bugs fixed by this?

I definitely had to add a few null checks in places to address warnings that came about through annotations. My guess is that as we start annotating other projects, we'll discover that the annotations here need to be tweaked as well. It's an iterative process in my experience.

> Is this introducing any runtime risks?

Sort of. In a few places I put `Assumes.NotNull`, as that teaches the compiler than something which it thinks may be null is not null after that point (because if it is null an exception is thrown.)

For example (simplified):

```c#
string? s = ...;
bool b = s is null;

if (!b)
{
    Assumes.NotNull(s);
    s = s.ToUpper();
}
```

Again this is simplified, but it shows a limit to the compiler's understanding of the code. It doesn't track that `b` tells us something about the null state of `s`.

So the risk here is whether I made a bad assumption in any of these cases. I was careful.

Note that `Assumes` throws exceptions which indicate that the bug is in the product, rather than user code.

Also note that in such a simple example as the above I'd likely just suppress with `!`, but in more complex examples it's not clear why the `!` is correct. We've been using NRT for over a year now on my team and we almost never use `!`, choosing to refactor code to be clearer. It's one of the things I like most about this language feature. However adding those refactorings to this PR is too risky and noisy. Better we improve over time. I'm happy to continue annotating the code over the coming weeks.

> How much of this was search&replace and how much required manual changes?

It was all manual and took several hours. You can see that the JetBrains annotations weren't used in all places. There were also some places where the JB annotations were incorrect (likely they fell out of date).

---

I've taken another look through these changes with a fresh pair of eyes and am happy with them. Any problems here would be easy to fix.

File: GitCommands/Git/GitModule.cs [hunk 1]
```diff
@@ -3367,25 +3352,27 @@ internal GitBlame ParseGitBlame(string output, Encoding encoding)
...
+                                committer!,
+                                committerMail!,
                                 committerTime,
-                                committerTimeZone,
-                                summary,
-                                filename);
-                            commitByObjectId[objectId] = commit;
+                                committerTimeZone!,
+                                summary!,
+                                filename!);
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs:3370]: This doesn't look good.

File: GitCommands/Git/GitItem.cs [hunk 2]
```diff
@@ -14,13 +16,13 @@ public GitItem(int mode, GitObjectType objectType, ObjectId objectId, string nam
             FileName = Name = name;
         }
 
-        public ObjectId ObjectId { get; }
+        public ObjectId? ObjectId { get; }
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitCommands/Git/GitItem.cs:19]: `IGitItem.ObjectId` is nullable, which is why this got annotated. In general, an implementation can be more strict (i.e. disallow nulls), so in this case I'll do that.

File: GitCommands/Remotes/ConfigFileRemoteSettingsManager.cs [hunk 3]
```diff
@@ -180,9 +180,9 @@ public IReadOnlyList<string> GetDisabledRemoteNames()
         {
             var module = GetModule();
             return module.LocalConfigFile.GetConfigSections()
-                .Where(s => s.SectionName == $"{DisabledSectionPrefix}remote")
+                .Where(s => s.SectionName == $"{DisabledSectionPrefix}remote" && s.SubSection is not null)
                 .Select(s => s.SubSection)
-                .ToList();
+                .ToList()!;
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitCommands/Remotes/ConfigFileRemoteSettingsManager.cs:185]: It's `List<string>` vs. `List<string?>`.

Unfortunately the compiler's doesn't interpret the null check two lines above. The easiest way was to `!` here.

I've changed this to introduce `LinqExtensions.WhereNotNull` with the appropriate annotations, so we can remove the `!`.

File: GitCommands/Utils/Strings.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Diagnostics.CodeAnalysis;
+
+#nullable enable
+
+namespace GitCommands.Utils
+{
+    // TODO remove this class when targeting .NET 5
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply GitCommands/Utils/Strings.cs:7]: Sure, so long as we don't expect to multi-target. I doubt Git Extensions would do so though.

I'll add:


```c#
#if NET5_0_OR_GREATER
#error Remove this class when targeting .NET 5 and update all usages
#endif
```

File: GitCommands/Git/GitModule.cs (hunk 1 above)
Comment (drewnoakes) [reply GitCommands/Git/GitModule.cs:3370]: Agree. This was probably the worst example of logic that tripped up NRT. Honestly I think it should be re-written to be clearer and have fewer moving parts.

**Comment by drewnoakes** [comment]
@RussKie I pushed changes to address your feedback. 

I also set `<Nullable>enable</Nullable>` at the project level and removed `#nullable enable` from all the source files. This means new code will be added with nullability by default, which is a better steady state.

**Review by RussKie — APPROVED** [review]

**Pull request merged**